feat(jobs): ValidateAction

sbliven opened the pull request:
## Description

Add ValidateAction to check Job DTO formatting

## Motivation

When creating a job clients can provide an untyped `jobParams` object in the DTO, or a `jobResultObject` when updating. These were previously not validated.



## Fixes:
This PR adds a new `validate` job action which can test for the presence of particular attributes in the DTO. Currently the attributes are specified using a [JSONPath](https://github.com/JSONPath-Plus/JSONPath) and are only verified to be non-null. If needed, future implementations could add JSONSchema support to do deeper validation.

For example, the following `jobConfig.json` section would configure the `public` job to require at least one datasetId:
```
{
      "jobType": "public",
      "create": {
        "auth": "#all",
        "actions": [
          {
            "actionType": "validate",
            "required": ["jobParams.datasetIds[0]"]
          }
        ]
      },
      "statusUpdate": {
        "auth": "#all"
      }
    }
```

## Changes:
- Implement `validate` action
- New dependency: jsonpath-plus
- Add tests
- Add example configuration

## Tests included

- [x] Included for each change/fix?
- [x] Passing? (Merge will not be approved unless this is checked) 

## Documentation
- [x] swagger documentation updated \[required\]
- [ ] official documentation updated \[nice-to-have]

sbliven committed:
Add JobAction.validate method.

This partially reverts commit b218952495b36d836c155e0ec0e7af9e3baf0832.

- Tightened bounds on DtoType for JobAction
- Validate DTOs

sbliven committed:
Remove job-create interceptor

DTO validation is in the controller, so the interceptor isn't needed.

sbliven committed:
Add jsonpath-plus dependency

sbliven committed:
Add ValidateAction

Also tests and example

sbliven commented:
What does everyone think of the use of JSONPath?

Pro: Simple syntax for simple reference, while keeping enough power for more complex structures like arrays
Pro: Well-known standard (eg used by `jq` command)
Con: Adds a dependency
Con: Only checks element presence, not type

A more powerful option would be to validate against a JSON Schema. This would allow type checking. However I think it would be much more verbose.

sofyalaski commented:
I think the JSONpath solution is appropriate for now. 
However, there is a minor issue in the jobs.controller: `performJobCreateAction` now accesses only the `jobInstance` of `JobClass` as an argument (this also needed to call the other function that performs the actions).  `performJobCreateAction` should take two arguments, the second being of `JobDto` type so that we can call the `validate` function. Now in the controller `validate` is called on `JobClass` instance, which is not how `validate` is defined.

sbliven commented:
Good catch, @sofyalaski. Why does this even compile?

sbliven committed:
Fix type error calling validate

sbliven commented:
Ah, it compiles because JobClass is structurally compatible with CreateJobDto. Working on a fix now.

sbliven committed:
Refactor job validate code

- Revert the previous commit, which was incorrect and incomplete
- Make valiateDTO and performActions generic so they can be shared
  between endpoints
- Separate out checkConfigVersion, which is specific to statusUpdate
- Move `getJobTypeConfiguration` calls up so this is only called once

sbliven commented:
This implementation has a couple problems.

First, the JSONPath solution doesn't perform any type checking. It would be nice to do something like
```
{
  "actionType": "validate",
  "required": {
    "jobParams.datasetIds": "object",
    "jobParams.datasetIds[*]": "string"
  }
}
```
This looks fine for simple types (string, number, list of literal values) but could quickly grow in complexity until it reaches a full type system. One solution would be to supply a JSON Schema to validate the DTO (which provides full type support). However including JSON schemas in the config file seems overly complex for operators. Is there a happy medium?

Second, the validate method currently acts against the DTO body. Half of the checks in v3 are actually against database entities referenced in the DTO or Job itself. For instance, "archive" jobs should check that all datasets associated with the job have "datasetLifecycle.archivable". Or "publish" jobs with file listings should check that all files exist, requiring checking associated DataBlocks from the dataset. Should these be added to a generic ValidateAction or be a more specific check? We are leaning towards handling those as special cases in the jobs controller.

sbliven committed:
Add mocha tests for ValidateAction.

- Add new test for validate jobs to check for required parameters
  in both POST and PATCH operations
- Remove some unused TestData which didn't match the current Job DTOs
- Add 'validate' job to the test jobconfig.json

sbliven committed:
Add type checking to ValidateAction

- rename top-level array to 'request'. This leaves open the possibility
  to validate other things beside the DTO
- now configuration takes the format of `"JSONPath": JSON-Schema`. The
  schema part is validated with Avj. Note that JSON Schemas can be very
  concise for simple datatypes, so this doesn't require deep knowledge
  of how to write JSONSchema.
- Update unit & mocha tests

sbliven committed:
Lint fixes

- remove 'any' via typeguard
- `npm run lint:fix`
- remove duplicate (and outdated) docstring

sbliven committed:
Merge branch 'release-jobs' into validateaction

sbliven committed:
Renumber tests within the 'Validate Job Action' section.

sbliven committed:
Merge branch 'release-jobs' into validateaction

(previous merge commit lost a parent)

sbliven commented:
I changed how this works somewhat. Now the ValidateAction config looks like:

```
{
  "actionType": "validate",
  "request": {
    "JSONPath": JSONSchema,
    ...
  }
}
```

The schema gets validated by avj, so you theoretically could use the full power of a schema. However combining it with the path allows most use cases to be covered by short, easily readable one-liners. Examples:

```
{
  "actionType": "validate",
  "request": {
    "jobParams.name": { "type": "string"}, // match simple types
    "jobParams.answers[*]": { "enum": ["yes","no"]}, // literal values (here applied to an array)
    "jobResultObject": {"$ref": "https://json.schemastore.org/schema-org-thing.json"}, // JSON Schema
  }
}
```

I spelled the top-level `request` to because it gets applied to the DTO (either the create or update DTO). If it needed to get applied somewhere else (eg to associated datasets) we can add that in the future.

sbliven commented on test/Jobs.js:
```diff
@@ -40,7 +39,7 @@ let datasetPid1 = null,
   encodedJobIdGroup4 = null,
   jobIdGroup5 = null,
-  encodedJobIdGroup5 = null,  
+  encodedJobIdGroup5 = null,
```
Sorry for the long diff. Everything down to like 4160 should be pure whitespaces changes (my editor removes trailing whitespace).

We should update eslint to cover js files to minimize these formatting changes.

sbliven commented on test/Jobs.js:
```diff
@@ -4158,4 +4156,192 @@ describe("1100: Jobs: Test New Job Model", () => {
       });
   });
+
+  describe("3120: Validate Job Action", () => {
```
This file is massive. I think it would be good to either split it into several files or at least group the tests with additional `describe` blocks.

Grouping them is nice for testing since you can just add `describe.only(` to run tests from a single section.

sbliven commented on test/TestData.js:
```diff
@@ -694,62 +694,6 @@ const TestData = {
     ],
   },
 
-  ArchiveJob: {
```
These weren't used by any tests. OK to delete them?

sbliven committed:
Fix jobconfig.example.json to match ValidateAction

sbliven commented on src/jobs/config/jobConfig.example.json:
```diff
@@ -58,7 +58,18 @@
+              "jobParams.datasetIds[*]": {
+                "type": "object",
+                "required": ["pid","files"]
+              }
```
Not the best example, as it's a bit complicated.

sbliven commented on src/jobs/config/jobConfig.updateExample.json:
```diff
@@ -18,7 +18,7 @@
           "actionType": "validate",
           "jobParamsSchema": {
-            "$schema": "https://json-schema.org/draft/2020-12/schema",
+            "$schema": "http://json-schema.org/draft-07/schema#",
```
One downside of Avj is that it needs to know the schema version during import. The default supports draft v6 and v7.

despadam commented on src/jobs/config/jobConfig.example.json:
```diff
@@ -58,7 +58,18 @@
+          {
+            "actionType": "validate",
+            "request": {
+              "jobParams.datasetIds[*]": {
```
Should now be `datasetList` instead of `datasetIds`

sbliven replied:
changed

despadam reviewed (changes requested):
Would it make more sense to try and also cover https://github.com/SciCatProject/scicat-backend-next/issues/1120 in this PR?

despadam commented:
Agreed to merge https://github.com/SciCatProject/scicat-backend-next/pull/1440, then remove `checkDatasetState` from the controller and incorporate its logic into the `validate` action, which will allow for configuring `jobType - datasetLifecycle` combinations and deleting the hardcoded `job-types.enum`. These changes will come in a new PR which will also incorporate this: https://github.com/SciCatProject/scicat-backend-next/issues/1120

sbliven committed:
Change example to use datasetList

sbliven committed:
Merge branch 'release-jobs' into validateaction

sbliven committed:
Whitespace changes

sbliven committed:
Merge branch 'release-jobs' into validateaction

sbliven committed:
Fix test value (incorrect following merge)

sbliven committed:
Fix failing test
